jax/_src/numpy/reductions.py

```diff
@@ -1810,16 +1810,19 @@ def _cumulative_reduction(
   if fill_nan:
     a = _where(lax_internal._isnan(a), _lax_const(a, fill_value), a)
 
+  a_type: DType = dtypes.dtype(a)
   result_type: DTypeLike = dtypes.dtype(dtype or a)
   if dtype is None and promote_integers or dtypes.issubdtype(result_type, np.bool_):
     result_type = _promote_integer_dtype(result_type)
-  result_type = dtypes.canonicalize_dtype(result_type)
+
+  if a_type != np.bool_ and dtype == np.bool_:
+    a = lax.ne(a, _lax_const(a, 0))
```

Comment:
If we use  `jnp.astype` in place of `lax.convert_element_type` here, it will be semantically equivalent to this new code.

Reply:
I see, though, there's both `dtype` and `result_type` in play, so this is probably cleaner.

Reply:
Done.

Reply:
Well, note I didn't *replace* the `convert_element_type`. I just changed this `ne`. Both conversions are needed; the `convert_element_type` is converting to an integer type, which has different semantics.